File: javascript_code/line/algo/sort/ThreeWayQuickSorter.js


// 时间复杂度：O(nlogn)
// 空间复杂度：O(1)
const sort = function (data) {
    if (data == null || data.length == 0) return

    const sortR = function (lo, hi) {
        if (lo >= hi) return
        // 分区
        let pivot = data[hi]

        let less = lo
        let great = hi

        let i = lo
        while (i <= great) {
            if (data[i] < pivot) {
                [data[less], data[i]] = [data[i], data[less]]
                less++
                i++
            } else if (data[i] > pivot) {
                [data[i], data[great]] = [data[great], data[i]]
                great--
            } else {
                i++
            }
        }
        // 对左边数组排序
        sortR(lo, less - 1) // log2n
        // 对右边数组排序
        sortR(great + 1, hi) // log2n
    }

    sortR(0, data.length - 1)
}



data = [34, 33, 12, 78, 21, 1, 98, 100]
sort(data)
console.log(data)

File: cpp_code/line/algo/sort/ThreeWayQuickSorter.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <algorithm>
#include <iostream>
#include <sstream>

using namespace std;

void sortR(int* data, int lo, int hi) {
    if (lo >= hi) return;
    // 分区
    int pivot = data[hi];

    int less = lo;
    int great = hi;

    int i = lo;
    while (i <= great) {
        if (data[i] < pivot) {
            swap(data[i], data[less]);
            less++;
            i++;
        } else if (data[i] > pivot) {
            swap(data[i], data[great]);
            great--;
        } else {
            i++;
        }
    }

    sortR(data, lo, less - 1);
    sortR(data, great +1, hi);
}


// 时间复杂度：O(nlogn)
// 空间复杂度：O(1)
void sort(int* data, int length) {
    if (data == nullptr || length <= 1) return;

    sortR(data, 0, length - 1);
}



int main() {
    int data[] = {12, 23, 36, 9, 24, 42};
    sort(data, 6);
    stringstream fmt;
    for (int i = 0; i < 6; ++i) {
        fmt << to_string(data[i]) << ",";
    }
    cout << fmt.str() << endl;
    return 0;
}